### PR TITLE
Implement adapter logic for Discord API Clients (discord.js, Eris)

### PR DESCRIPTION
**Pull request by mariusbegby**
# 🚨 Work in progress 🚨

## Changes
As `discord-player` want to implement support for Eris and potentially other Discord API client libraries, we see the need to restructure some of our current code. Previously all logic has been written with discord.js in mind, and also many places directly depend on discord.js.

This PR aims to move all logic depending on a Discord API client today inside an adapter. The idea is that we have an adapter interface which defines methods that can be used throughout the codebase where we today use a Discord API client, such as for resolving a voice channel by id, or getting information about a guild member. This means the codebase can use these common methods, without needing to know about the underlying implementation, such as if it's using `discord.js` or `Eris`.

This PR is still a work in progress, and it's not yet clear if implementation details will change.

## Status

- [ ] These changes have been tested and formatted properly.
  - These changes have been somewhat tested, but needs more extensive testing. When released should first be released on a non-stable release channel for more large-scale user testing and feedback.
- [ ] This PR includes only documentation changes, no code change.
- [x] This PR introduces some Breaking changes. 
  - 🚨 Some public method signatures have changed. Now only taking in `id` instead of also supporting full discord.js specific object such as user, channel, etc. That means for `Player.play()` you cannot pass in a full discord.js channel object, and instead need to pass in the channel id, which is an universal identifier, aka `Snowflake`.

### CODE REVIEW

File: packages/discord-player/src/clientadapter/IClientAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import { AudioPlayerState, DiscordGatewayAdapterCreator } from "discord-voip";
+
+export const ClientType = {
+    DiscordJS: "discord.js",
+    Eris: "eris",
+    Unknown: "unknown"
+};
+
+export enum ChannelType {
```
Comment (twlite) [inline packages/discord-player/src/clientadapter/IClientAdapter.ts:9]: this is given by `discord-api-types` package

**Review by twlite — CHANGES_REQUESTED** [review]

File: packages/discord-player/src/clientadapter/IClientAdapter.ts (hunk 1 above)
Comment (mariusbegby) [reply packages/discord-player/src/clientadapter/IClientAdapter.ts:9]: Thanks, will check it out.

Comment (mariusbegby) [reply packages/discord-player/src/clientadapter/IClientAdapter.ts]: Resolved.

File: packages/discord-player/src/DefaultVoiceStateHandler.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import { Snowflake } from 'discord-api-types/globals';
```
Comment (retrouser955) [inline packages/discord-player/src/DefaultVoiceStateHandler.ts:1]: Could you change the import to import a type instead?

File: packages/discord-player/src/clientadapter/IClientAdapter.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { ChannelType } from 'discord-api-types/v9';
```
Comment (twlite) [inline packages/discord-player/src/clientadapter/IClientAdapter.ts]: lets use v10 🙂 
```suggestion
import { ChannelType } from 'discord-api-types/v10';
```

**Review by twlite — CHANGES_REQUESTED** [review]

File: packages/discord-player/src/clientadapter/IClientAdapter.ts (hunk 3 above)
Comment (retrouser955) [reply packages/discord-player/src/clientadapter/IClientAdapter.ts]: Is this outdated? Seems latest uses v10

File: packages/discord-player/src/clientadapter/IClientAdapter.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import { AudioPlayerState, DiscordGatewayAdapterCreator } from "discord-voip";
+
+export const ClientType = {
+    DiscordJS: "discord.js",
+    Eris: "eris",
+    Unknown: "unknown"
+};
+
+export type ClientUser = {
+    setSelfDeaf(state?: boolean, reason?: string): Promise<void>;
```
Comment (retrouser955) [inline packages/discord-player/src/clientadapter/IClientAdapter.ts:12]: Can we set these state to a non undefined property?
```suggestion
    setSelfDeaf(state: boolean, reason?: string): Promise<void>;
```

Comment (retrouser955) [reply packages/discord-player/src/clientadapter/IClientAdapter.ts:12]: Same for other boolean logic like setSelfDeaf

File: packages/discord-player/src/clientadapter/IClientAdapter.ts (hunk 3 above)
Comment (mariusbegby) [reply packages/discord-player/src/clientadapter/IClientAdapter.ts]: I think you reviewed outdated code as I already updated all to v10. This is resolved. I agree we should use v10.

File: packages/discord-player/src/clientadapter/IClientAdapter.ts (hunk 4 above)
Comment (mariusbegby) [reply packages/discord-player/src/clientadapter/IClientAdapter.ts]: Yes but our public methods for GuildQueue.setSelfDeaf() and GuildQueue.setSelfMute() takes in mode?: boolean

I preserved this initially but as we have breaking changes anyways I will change this.

Resolved.

**Pull request closed without merging**